File: aoj/src/ALDS1_4_D/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const maxWeight = 10000

func main() {
	stdin := bufio.NewScanner(os.Stdin)
	stdin.Scan()
	t := stdin.Text()
	//n: 荷物数, k: トラック数
	n, k := parseCondition(t)

	// w: n個の荷物の重さ
	w := make([]int, n, n)
	i := 0
	for stdin.Scan() {
		t := stdin.Text()
		wi, err := strconv.Atoi(t)
		if err != nil {
			panic(err)
		}
		w[i] = wi
		i++
	}

	ii := binarySearch(maxWeight*n, func(i int) bool {
		//fmt.Println(w, k, i)
		return canBeAllocated(w, k, i)

	})
	fmt.Println(ii)
}

func binarySearch(right int, ORHigher func(int) bool) int {
	left := 0
	for left < right {

		mid := (right + left) / 2
		//fmt.Println(left, right, mid)
		if ORHigher(mid) {
			right = mid
		} else {
			left = mid + 1
		}
	}
	return right
}

func parseCondition(s string) (int, int) {
	ss := strings.Split(s, " ")
	n, err := strconv.Atoi(ss[0])
	if err != nil {
		panic(err)
	}

	k, err := strconv.Atoi(ss[1])
	if err != nil {
		panic(err)
	}

	return n, k
}

func canBeAllocated(w []int, k, p int) bool {
	load := 0
	allocated := 0
	for _, wi := range w {
		//fmt.Printf("load: %v, allocated: %v, wi: %v, k: %v, p: %v\n", load, allocated, wi, k, p)
		if wi > p {
			return false
		}
		if load+wi > p {
			allocated++
			if allocated > k {
				return false
			}
			load = 0
		}
		load += wi
	}
	if allocated >= k {
		return false
	}
	return true
}

File: aoj/src/ALDS1_4_D/main.py
# メイン関数
# 比較関数
# バイナリサーチ
from sys import stdin
import math

MAX_WEIGHT = 10_000

def carriable(k, max, wi):
    shipped = 0
    picked = 0
    for w in wi:
        if picked + w > max:
            picked = 0
            shipped +=1
        picked += w
        if shipped > k or w > max:
            return False
    shipped += 1
    return shipped <= k


def binarySearch(k, max, wi):
    right = max
    left = 0
    while left < right:
        mid = int((left + right) /2 )
        if carriable(k, mid, wi):
            right = mid
        else:
            left = mid + 1
    return right

def main():
    args = stdin.readline().strip().split(" ")
    n = int(args[0])
    k = int(args[1])
    wi = [int(input()) for _ in range(0, n)]
    max = math.ceil(n * MAX_WEIGHT / k)
    result = binarySearch(k, max, wi)
    print(result)

if __name__ == '__main__':
    main()
